### ISSUE TITLE
accessing the highchart instance

### ISSUE THREAD

**Issue by msanjdev** [highcharts/highcharts-angular#11]
(Thank you for the wrapper)

We've got the hello world app working fine in our angular 5 app. We want to do some computations and update the options. I see that I can update this.chartOptions to the new values but I don't see the graph redrawing. Somewhere on the net / docs I see that we may need to call the redraw function of the api. We need the chart instance even to do things like addSeries etc.

I tried adding a chart: object definition but that didn't seem to help. Did go through [instance details section](https://github.com/highcharts/highcharts-angular#highcharts-instance-details) I seem to have missed getting the specific chart instance.

Any suggestions would be welcome - thanks.

**Comment by KacperMadej** [comment]
Hi @msanjdev,

Highcharts instance is loaded Highcharts and specific chart can be accessed through `Highcharts.charts` array - it holds all charts created using that Highcharts instance. Single chart is not accessible through the components interface, but it's [here](https://github.com/highcharts/highcharts-angular/blob/master/src/app/highcharts-chart.component.ts#L11). You can access 'chart' e.g. through [chart's events](https://api.highcharts.com/highcharts/chart.events.redraw) - `this` is the `chart` there, or callback - 4th in [here](https://github.com/highcharts/highcharts-angular#options-details).

The repo holds a demo, so I'll reference code [here](https://github.com/highcharts/highcharts-angular/tree/master/src/app) to better explain:
* If you update chart's options by altering the options object enough to trigger Angular default change checker like [here](https://github.com/highcharts/highcharts-angular/blob/master/src/app/app.component.ts#L49-L51). New object is created, so that's a big enough change to be detected automatically.

* If you change only options that are deeper in the options object structure, then you will need to set `updateFlag` (number 5 in [here](https://github.com/highcharts/highcharts-angular#options-details)) to `true`. After the change will be applied `updateFlag` will [change back to false](https://github.com/highcharts/highcharts-angular/blob/master/src/app/highcharts-chart.component.ts#L26). It's demonstrated in [this code](https://github.com/highcharts/highcharts-angular/blob/master/src/app/app.component.ts#L75-L83) - the `updateFlag` is defined [here](https://github.com/highcharts/highcharts-angular/blob/master/src/app/app.component.ts#L70), changed [here](https://github.com/highcharts/highcharts-angular/blob/master/src/app/app.component.ts#L82) and the variable is bound [here](https://github.com/highcharts/highcharts-angular/blob/master/src/app/app.component.html#L47).

A lot of "here"s and "there"s, but all are linked.

If anything is unclear, please let me know. If you have more technical questions you can contact Highcharts support - https://www.highcharts.com/support

**Comment by msanjdev** [comment]
yeah I see how they're all linked - thanks I accessed the instance using the chartCallback event and stored it as a class variable.

I wanted to redraw the chart on a response to a this.http.get(url).subscribe(data => {
// redraw
})

but finally got to do it after a long time & effort :)

**Comment by KacperMadej** [comment]
@msanjdev : Happy to hear it works in the end.

If you could provide more info about what was the most unclear or misleading, what caused the problem, then maybe we could create some FAQ or trouble shooting.

**Comment by Heli515** [comment]
Hi, I'm having a very similar problem. I'm playing with Demo#2 and hoping to add/delete series on the chart. The solution I can think of now is to push the new data to 'series'. However, chart is not updating even setting the flag to true. Do you mind provide how you solved the problem? Thanks!

**Comment by msanjdev** [comment]
Sorry I missed @KacperMadej 's message earlier and would've elaborated with a little more detail in my response (will take it as a learning in future). 

This is the best I can recollect now:

As far as I remember,  the root cause here was due to a _copy of the instance_ on which we're calling a function with data and wondering why it isn't updating, versus a reference to the **actual instance**. 

I had meant the solution had been finally to simply to store the **actual instance** as a class variable, as he had suggested.

I used what @KacperMadej described: initialize the class variable using one of the chartCallback events. 

Please look at the links that he refers to in his line above:

> You can access 'chart' e.g. through [chart's events](https://api.highcharts.com/highcharts/chart.events.redraw) - this is the chart there, or callback - 4th in [here](https://github.com/highcharts/highcharts-angular#options-details).

In such situations, I usually use a browser inspector like Firebug or Chrome Inspector to set a breakpoint in as many events as possible. The event handler could just contain a single line placeholder of a console.log line, where I can set the breakpoint, and when paused inspect which event parameter has got a workable instance. I may even use the Console of the Inspector to run a line of code with that instance to see if the graph updates with the updated data.

**Comment by Heli515** [comment]
@Chrisk12 Hi Chris, I created a class variable in my component and assigned the callback to it

**Comment by xts-velkumars** [comment]
@msanjdev 

Could you share your code? 

I have the scenario, where i need to draw a chart inside a ngFor angular. My html looks below. I have chart data (XAxis and YAxis inside timeSheet object)
really appreciated, if any one help me 

```
<div *ngFor="let timeSheet of  timeSheetList">            
        <div fxLayout="row" class="mar-top-20 apr-chart-container">
                      <highcharts-chart [Highcharts]="highcharts"
                                        [constructorType]="chartConstructor"
                                        [options]="chartOptions"
                                        [callbackFunction]="chartCallback(highcharts, timeSheet)"
                                        [(update)]="updateFlag"
                                        [oneToOne]="oneToOneFlag"
                                        [runOutsideAngular]="runOutsideAngularFlag"
                                        style="width: 480px; height: 120px; display: block;"></highcharts-chart>
      </div>
</div>
```
